nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala

```diff
@@ -31,381 +36,698 @@ sealed abstract class Array[T]
 }
 
-object BooleanArray {
-  def copy(from: BooleanArray, fromPos: Int,
-           to: BooleanArray, toPos: Int, length: Int): Unit = {
-    ???
-  }
+// ###sourceLocation(file: "/home/francois/proyectos/oss/scala-native-fbd/scala-native/nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala.gyb", line: 210)
 
-  def alloc(length: Int): BooleanArray = {
-    val arrty = typeof[BooleanArray]
-    val arrsize = sizeof[ArrayHeader] + sizeof[Boolean] * length
-    val arr = runtime.allocAtomic(arrty, arrsize)
-    // set the length
-    !(arr.cast[Ptr[Byte]] + sizeof[Ptr[_]]).cast[Ptr[Int]] = length
-    arr.cast[BooleanArray]
-  }
-}
+final class BooleanArray private () extends PrimitiveArray[Boolean] {
+  def apply(i: Int): Boolean = ! (pointerAt(sizeof[Boolean], i).cast[Ptr[Boolean]])
 
-// ###sourceLocation(file: "/Users/Denys/.src/native/nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala.gyb", line: 79)
+  def update(i: Int, value: Boolean): Unit = ! (pointerAt(sizeof[Boolean], i).cast[Ptr[Boolean]]) = value
 
-final class CharArray private () extends Array[Char] {
-  def apply(i: Int): Char =
-    if (i < 0 || i >= length)
-      throw new IndexOutOfBoundsException(i.toString)
-    else {
-      val headptr = (this.cast[Ptr[Byte]] + sizeof[ArrayHeader]).cast[Ptr[Char]]
-      headptr(i)
-    }
+  protected override def clone(): BooleanArray = helperClone (sizeof[Boolean]).cast[BooleanArray]
+  
+  protected def copyTo (fromPos: Int, toPos: Int): Unit = update (toPos, apply(fromPos)) 
 
-  def update(i: Int, value: Char): Unit =
-    if (i < 0 || i >= length)
-      throw new IndexOutOfBoundsException(i.toString)
-    else {
-      val headptr = (this.cast[Ptr[Byte]] + sizeof[ArrayHeader]).cast[Ptr[Char]]
-      headptr(i) = value
+  def sizeofElement: CSize = sizeof[Boolean]
+
+  def copyToArray (fromPos: Int, to : BooleanArray, toPos: Int, len: Int): Unit = { 
+    // methods for overlap copy (note: this == to) 
+    def forwardCopy() {
+      var c = 0
+           
+      while (c < len) {
+        copyTo (fromPos + c, toPos + c)
+        c += 1
+      }
+    }
+    
+    def backwardCopy() {
+      var c = len - 1
+           
+      while (c >= 0) {
+        copyTo (fromPos + c, toPos + c)        
+        c -= 1
+      }
     }
+    
+    // sanity check
+    if (to == null)
+      throw new NullPointerException()
+    
+    val overlapDirection = Array.validateBoundaries (this, fromPos, to, toPos, len)
+    
+    val fromPtr: Ptr[Byte] = (this.cast[Ptr[Byte]] + sizeof[ArrayHeader] + sizeofElement * fromPos).cast[Ptr[Byte]]
+    
+    val toPtr: Ptr[Byte] = (to.cast[Ptr[Byte]] + sizeof[ArrayHeader] + sizeofElement * toPos).cast[Ptr[Byte]]
+    
+    if (overlapDirection == 0)
+      // no overlap, use LLVM memcpy
+      `llvm.memcpy.p0i8.p0i8.i64`(toPtr, fromPtr, sizeofElement * len, 1, false)
+    else if (overlapDirection > 0)
+      forwardCopy()
+    else
+      backwardCopy()
```

Comment:
Can't we just use `memmove` here? That's what apache harmony does for copying (you can find their implementation in `array_copy` function that is located in `vm_arrays.cpp`)